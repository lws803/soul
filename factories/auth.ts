import { Factory } from 'fishery';

import { RefreshToken } from 'src/auth/entities/refresh-token.entity';
import { JWTRefreshPayload } from 'src/auth/entities/jwt-refresh-payload.entity';
import { JWTPayload } from 'src/auth/entities/jwt-payload.entity';
import { TokenType } from 'src/auth/enums/token-type.enum';

import * as factories from './index';

export const refreshToken = Factory.define<RefreshToken>(() => ({
  id: 1,
  user: factories.user.build(),
  isRevoked: false,
  createdAt: new Date('1995-12-17T03:24:00'),
  updatedAt: new Date('1995-12-18T03:24:00'),
  expires: new Date('1995-12-19T03:24:00'),
  platformUser: factories.platformUser.build(),
}));

export const jwtRefreshPayload = Factory.define<JWTRefreshPayload>(() => ({
  tokenId: refreshToken.build().id,
  userId: factories.user.build().id,
  tokenType: TokenType.Refresh,
}));

export const jwtPayload = Factory.define<JWTPayload>(() => {
  const oneUser = factories.user.build();
  return {
    userId: oneUser.id,
    username: oneUser.username,
    tokenType: TokenType.Access,
  };
});

export const jwtPayloadWithPlatform = Factory.define<JWTPayload>(() => {
  const oneUser = factories.user.build();
  const onePlatformUser = factories.platformUser.build({ user: oneUser });
  return {
    userId: oneUser.id,
    username: oneUser.username,
    tokenType: TokenType.Access,
    platformId: onePlatformUser.platform.id,
    roles: onePlatformUser.roles,
  };
});

export const jwtRefreshPayloadWithPlatform = Factory.define<JWTRefreshPayload>(
  () => {
    const oneUser = factories.user.build();
    const onePlatformUser = factories.platformUser.build({ user: oneUser });
    return {
      tokenId: refreshToken.build().id,
      userId: oneUser.id,
      tokenType: TokenType.Refresh,
      platformId: onePlatformUser.platform.id,
      roles: onePlatformUser.roles,
    };
  },
);

export const requestUserObject = Factory.define<{
  tokenType: TokenType;
  username: string;
  userId: number;
}>(() => ({
  tokenType: TokenType.Access,
  username: 'TEST_USER',
  userId: 1,
}));
