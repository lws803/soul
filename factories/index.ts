export * from './users';
export * from './auth';
export * from './platform';
export * from './user-connections';
