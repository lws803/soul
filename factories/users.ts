import { Factory } from 'fishery';

import { CreateUserDto, UpdateUserDto } from 'src/users/serializers/api.dto';
import { User } from 'src/users/entities/user.entity';

export const user = Factory.define<User>(({ sequence }) => {
  // Rewinds the sequence as we do not want to have "sequences" bleed into subsequent builds
  user.rewindSequence();
  return {
    id: sequence,
    username: `test-user-${sequence}`,
    userHandle: `test-user-${sequence}#${sequence}`,
    email: `TEST_USER_${sequence}@EMAIL.COM`,
    isActive: true,
    createdAt: new Date('1995-12-17T03:24:00'),
    updatedAt: new Date('1995-12-18T03:24:00'),
    hashedPassword: 'TEST_HASHED_PASSWORD',
    displayName: 'DISPLAY_NAME',
    bio: 'BIO',
  };
});

export const createUserDto = Factory.define<CreateUserDto>(() => ({
  email: 'TEST_USER@EMAIL.COM',
  username: 'test-user',
  password: 'TEST_PASSWORD',
  bio: null,
  displayName: null,
}));

export const updateUserDto = Factory.define<UpdateUserDto>(() => ({
  email: 'UPDATED_EMAIL@EMAIL.COM',
  username: 'updated-user',
  displayName: 'UPDATED_DISPLAY_NAME',
  bio: 'UPDATED_BIO',
}));
