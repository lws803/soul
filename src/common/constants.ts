export const DEFAULT_NUM_ITEMS_PER_PAGE = 10;
