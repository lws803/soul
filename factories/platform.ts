import { Factory } from 'fishery';

import * as factories from 'factories';
import { Platform } from 'src/platforms/entities/platform.entity';
import { PlatformUser } from 'src/platforms/entities/platform-user.entity';
import {
  UpdatePlatformDto,
  CreatePlatformDto,
} from 'src/platforms/dto/api.dto';
import { UserRole } from 'src/roles/role.enum';
import { PlatformCategory } from 'src/platforms/entities/platform-category.entity';

export const onePlatformCategory = Factory.define<PlatformCategory>(() => ({
  id: 1,
  name: 'CATEGORY',
}));

export const onePlatform = Factory.define<Platform>(() => ({
  id: 1,
  name: 'TEST_PLATFORM',
  nameHandle: 'test_platform#1',
  createdAt: new Date('1995-12-17T03:24:00'),
  updatedAt: new Date('1995-12-18T03:24:00'),
  userConnections: [],
  isVerified: true,
  redirectUris: ['TEST_REDIRECT_URI'],
  category: onePlatformCategory.build(),
  activityWebhookUri: 'ACTIVITY_WEBHOOK_URI',
}));

export const platformArray = Factory.define<Platform[]>(() => [
  onePlatform.build(),
  onePlatform.build({ id: 2, name: 'TEST_PLATFORM_2' }),
]);

export const onePlatformUser = Factory.define<PlatformUser>(() => ({
  id: 1,
  user: factories.oneUser.build(),
  platform: factories.onePlatform.build(),
  roles: [UserRole.Admin, UserRole.Member],
  createdAt: new Date('1995-12-17T03:24:00'),
  updatedAt: new Date('1995-12-18T03:24:00'),
}));

export const platformUserArray = Factory.define<PlatformUser[]>(() => [
  onePlatformUser.build(),
  onePlatformUser.build({ id: 2 }),
]);

export const updatePlatformDto = Factory.define<UpdatePlatformDto>(() => ({
  name: 'TEST_PLATFORM_UPDATE',
  category: 'CATEGORY_UPDATE',
}));

export const createPlatformDto = Factory.define<CreatePlatformDto>(() => ({
  name: 'TEST_PLATFORM',
  redirectUris: ['TEST_REDIRECT_URI'],
  category: onePlatformCategory.build().name,
}));

type CreatePlatformRequest = {
  name: string;
  redirect_uris: string[];
  category: string;
};

export const createPlatformRequestDto = Factory.define<CreatePlatformRequest>(
  () => ({
    name: 'TEST_PLATFORM',
    redirect_uris: ['TEST_REDIRECT_URI'],
    category: onePlatformCategory.build().name,
  }),
);
