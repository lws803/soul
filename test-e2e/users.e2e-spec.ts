import { HttpStatus, INestApplication } from '@nestjs/common';
import { Connection, Repository } from 'typeorm';
import * as request from 'supertest';

import { User } from 'src/users/entities/user.entity';

import * as factories from '../factories';

import createAppFixture from './fixtures/create-app-fixture';
import {
  createUsersAndLoginFixture,
  UserAccount,
} from './fixtures/create-users-and-login-fixture';

describe('UsersController (e2e)', () => {
  let app: INestApplication;
  let connection: Connection;
  let userRepository: Repository<User>;

  beforeAll(async () => {
    app = await createAppFixture({});
    await app.init();
    app.useLogger(false);

    connection = app.get(Connection);
    userRepository = connection.getRepository(User);
    await connection.synchronize(true);
  });

  afterAll(async () => {
    await app.close();
  });

  describe('/users (POST)', () => {
    afterEach(async () => {
      await userRepository.delete({});
    });

    it('creates a new user successfully', () => {
      return request(app.getHttpServer())
        .post('/users')
        .send(
          factories.createUserDto.build({
            password: '3Yarw#Nm%cpY9QV&',
          }),
        )
        .expect(HttpStatus.CREATED)
        .expect((res) => {
          expect(res.body).toStrictEqual({
            id: 1,
            email: 'TEST_USER@EMAIL.COM',
            username: 'TEST_USER',
            user_handle: 'test_user#1',
            is_active: false,
            created_at: expect.any(String),
            updated_at: expect.any(String),
          });
        });
    });

    it('throws user duplicate error', async () => {
      await userRepository.save(factories.user.build());
      return request(app.getHttpServer())
        .post('/users')
        .send(factories.createUserDto.build({ password: '3Yarw#Nm%cpY9QV&' }))
        .expect(HttpStatus.CONFLICT)
        .expect((res) => {
          expect(res.body).toStrictEqual({
            error: 'DUPLICATE_USER_EXISTS',
            message:
              'A user with the email address: TEST_USER@EMAIL.COM already exists. ' +
              'Please login or use a different email address.',
          });
        });
    });
  });

  describe('/users (GET)', () => {
    beforeAll(async () => {
      await userRepository.save([
        factories.user.build({ id: undefined }),
        factories.user.build({
          id: undefined,
          username: 'TEST_USER_2',
          userHandle: 'test_user_2#2',
          email: 'TEST_USER_2@EMAIL.COM',
        }),
      ]);
    });

    afterAll(async () => {
      await userRepository.delete({});
    });

    it('should return full list of users', async () => {
      return request(app.getHttpServer())
        .get('/users')
        .expect(HttpStatus.OK)
        .expect((res) => {
          expect(res.body).toStrictEqual({
            total_count: 2,
            users: [
              {
                id: expect.any(Number),
                user_handle: 'test_user_2#2',
                username: 'TEST_USER_2',
              },
              {
                id: expect.any(Number),
                user_handle: 'test_user#1',
                username: 'TEST_USER',
              },
            ],
          });
        });
    });

    it('should return partial list of users with pagination', async () => {
      return request(app.getHttpServer())
        .get('/users?page=1&num_items_per_page=1')
        .expect(HttpStatus.OK)
        .expect((res) => {
          expect(res.body).toStrictEqual({
            total_count: 2,
            users: [
              {
                id: expect.any(Number),
                user_handle: 'test_user_2#2',
                username: 'TEST_USER_2',
              },
            ],
          });
        });
    });

    it('should return partial list of users with full text search', async () => {
      return request(app.getHttpServer())
        .get('/users?q=TEST_USER_2')
        .expect(HttpStatus.OK)
        .expect((res) => {
          expect(res.body).toStrictEqual({
            total_count: 1,
            users: [
              {
                id: expect.any(Number),
                user_handle: 'test_user_2#2',
                username: 'TEST_USER_2',
              },
            ],
          });
        });
    });
  });

  describe('/users/:id (GET)', () => {
    beforeAll(async () => {
      await userRepository.save(factories.user.build());
    });

    afterAll(async () => {
      await userRepository.delete({});
    });

    it('should return a user', async () => {
      return request(app.getHttpServer())
        .get('/users/1')
        .expect(HttpStatus.OK)
        .expect((res) => {
          expect(res.body).toStrictEqual({
            id: expect.any(Number),
            user_handle: 'test_user#1',
            username: 'TEST_USER',
          });
        });
    });

    it('should return user not found', async () => {
      return request(app.getHttpServer())
        .get('/users/999')
        .expect(HttpStatus.NOT_FOUND)
        .expect((res) => {
          expect(res.body).toStrictEqual({
            error: 'USER_NOT_FOUND',
            message:
              'A user with the id: 999 could not be found, please try again.',
          });
        });
    });
  });

  describe('/users/:id (PATCH)', () => {
    let userAccount: UserAccount;

    beforeEach(async () => {
      const [firstUser] = await createUsersAndLoginFixture(app);
      userAccount = firstUser;
    });

    afterEach(async () => {
      await userRepository.delete({});
    });

    it('should update myself', async () => {
      return request(app.getHttpServer())
        .patch('/users/me')
        .set('Authorization', `Bearer ${userAccount.accessToken}`)
        .send(factories.updateUserDto.build())
        .expect(HttpStatus.OK)
        .expect((res) => {
          expect(res.body).toStrictEqual({
            id: userAccount.user.id,
            user_handle: `updated_user#${userAccount.user.id}`,
            username: 'UPDATED_USER',
            email: 'UPDATED_EMAIL@EMAIL.COM',
            is_active: true,
            created_at: expect.any(String),
            updated_at: expect.any(String),
          });
        });
    });
  });

  describe('/users/:id (DELETE)', () => {
    let userAccount: UserAccount;

    beforeEach(async () => {
      const [firstUser] = await createUsersAndLoginFixture(app);
      userAccount = firstUser;
    });

    afterEach(async () => {
      await userRepository.delete({});
    });

    it('should delete myself', async () => {
      return request(app.getHttpServer())
        .delete('/users/me')
        .set('Authorization', `Bearer ${userAccount.accessToken}`)
        .expect(HttpStatus.OK)
        .expect((res) => {
          expect(res.body).toStrictEqual({});
        });
    });
  });

  describe('/users/me (GET)', () => {
    let firstUserAccessToken;

    beforeAll(async () => {
      const [{ accessToken }] = await createUsersAndLoginFixture(app);
      firstUserAccessToken = accessToken;
    });

    afterAll(async () => {
      await userRepository.delete({});
    });

    it('retrieves me', async () => {
      return request(app.getHttpServer())
        .get('/users/me')
        .set('Authorization', `Bearer ${firstUserAccessToken}`)
        .expect(HttpStatus.OK)
        .expect((res) =>
          expect(res.body).toStrictEqual({
            created_at: expect.any(String),
            updated_at: expect.any(String),
            id: expect.any(Number),
            email: 'TEST_USER@EMAIL.COM',
            is_active: true,
            user_handle: expect.any(String),
            username: 'TEST_USER',
          }),
        );
    });
  });
});
