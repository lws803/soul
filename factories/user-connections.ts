import { Factory } from 'fishery';

import { FindOneUserConnectionResponseEntity } from 'src/user-connections/serializers/api-responses.entity';

import { UserConnection } from 'src/user-connections/entities/user-connection.entity';

import { platformEntity } from './platform';
import { userEntity } from './users';

export const userConnectionEntity = Factory.define<UserConnection>(() => ({
  id: 1,
  fromUser: userEntity.build(),
  toUser: userEntity.build({ id: 2, email: 'TEST_USER_2@EMAIL.COM' }),
  platforms: [],
  createdAt: new Date('1995-12-17T03:24:00'),
  updatedAt: new Date('1995-12-18T03:24:00'),
  mutualConnection: null,
}));

export const oneUserConnectionResponseEntity =
  Factory.define<FindOneUserConnectionResponseEntity>(() => ({
    id: 1,
    fromUser: userEntity.build(),
    toUser: userEntity.build({ id: 2, email: 'TEST_USER_2@EMAIL.COM' }),
    platforms: [],
    createdAt: new Date('1995-12-17T03:24:00'),
    updatedAt: new Date('1995-12-18T03:24:00'),
    mutualConnection: null,
    isMutual: false,
  }));

export const userConnectionEntityArray = Factory.define<UserConnection[]>(
  () => [
    userConnectionEntity.build({ id: 1 }),
    userConnectionEntity.build({
      id: 2,
      fromUser: userEntity.build({ id: 3 }),
      toUser: userEntity.build({ id: 4 }),
    }),
  ],
);

type CreateUserConnectionRequest = {
  to_user_id: number;
  platform_id?: number;
};

export const createUserConnectionRequest =
  Factory.define<CreateUserConnectionRequest>(() => ({
    to_user_id: userEntity.build({ id: 2 }).id,
  }));

type PostPlatformToUserConnectionRequest = {
  platform_id: number;
};

export const postPlatformToUserConnectionRequest =
  Factory.define<PostPlatformToUserConnectionRequest>(() => ({
    platform_id: platformEntity.build().id,
  }));
