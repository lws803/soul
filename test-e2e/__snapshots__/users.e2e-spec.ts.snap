// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersController (e2e) /users (GET) should return full list of users 1`] = `
Object {
  "total_count": 2,
  "users": Array [
    Object {
      "id": 4,
      "user_handle": "test_user_2#2",
      "username": "TEST_USER_2",
    },
    Object {
      "id": 3,
      "user_handle": "test_user_1#1",
      "username": "TEST_USER_1",
    },
  ],
}
`;

exports[`UsersController (e2e) /users (GET) should return partial list of users with full text search 1`] = `
Object {
  "total_count": 1,
  "users": Array [
    Object {
      "id": 4,
      "user_handle": "test_user_2#2",
      "username": "TEST_USER_2",
    },
  ],
}
`;

exports[`UsersController (e2e) /users (GET) should return partial list of users with pagination 1`] = `
Object {
  "total_count": 2,
  "users": Array [
    Object {
      "id": 4,
      "user_handle": "test_user_2#2",
      "username": "TEST_USER_2",
    },
  ],
}
`;

exports[`UsersController (e2e) /users (POST) creates a new user successfully 1`] = `
Object {
  "created_at": Any<String>,
  "email": "TEST_USER@EMAIL.COM",
  "id": 1,
  "is_active": false,
  "updated_at": Any<String>,
  "user_handle": "test_user#1",
  "username": "TEST_USER",
}
`;

exports[`UsersController (e2e) /users/:id (GET) should return a user 1`] = `
Object {
  "id": 1,
  "user_handle": "test_user_1#1",
  "username": "TEST_USER_1",
}
`;

exports[`UsersController (e2e) /users/:id (PATCH) should update myself 1`] = `
Object {
  "created_at": Any<String>,
  "email": "UPDATED_EMAIL@EMAIL.COM",
  "id": 5,
  "is_active": true,
  "updated_at": Any<String>,
  "user_handle": "updated_user#5",
  "username": "UPDATED_USER",
}
`;

exports[`UsersController (e2e) /users/me (GET) retrieves me 1`] = `
Object {
  "created_at": Any<String>,
  "email": "TEST_USER@EMAIL.COM",
  "id": Any<Number>,
  "is_active": true,
  "updated_at": Any<String>,
  "user_handle": Any<String>,
  "username": "TEST_USER",
}
`;
